docs/dev/specs/packable.md

```diff
@@ -0,0 +1,284 @@
+The proposed trait was the following
+
+```rust
+pub trait Packable {
+    type Error: Debug;
+
+    fn packed_len(&self) -> usize;
+
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error>;
+
+    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
+```
+
+The main issue with this trait is that it cannot be used in a `no_std`
+environment because it depends explicitly on the `std::io` API, whose
+transition to the `core` crate has not been decided yet.
+
+Another issue is that the `Error` type is used to represent three different
+kinds of errors:
+
+- Serialization: Raised when there are issues while writing bytes.
+- Deserialization errors: Raised when there are issues while reading bytes.
+- Semantic errors: Raised when the bytes being used to create a value are
+  invalid for the data layout of such value.
+
+# Design
+
+## Replacing `std::io`
+
+We introduce two new traits to replace `Read` and `Write` from `std::io`.
+
+First we have the `Unpacker` trait which represents a type that can be used to
+read bytes from it.
+
+```rust
+pub trait Unpacker: Sized {
+    type Error;
+
+    fn unpack_bytes(&mut self, slice: &mut [u8]) -> Result<(), Self::Error>;
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  deserialization.
+
+- The `unpack_bytes` method writes enough bytes to fill `slice`. This method
+  must fail if the unpacker does not have enough bytes to fulfill the request.
+
+We also have the `Packer` trait which represents a type that can be used to
+write bytes into it.
+
+```rust
+pub trait Packer {
+    type Error;
+
+    fn pack_bytes(&mut self, bytes: &[u8]) -> Result<(), Self::Error>;
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  serialization.
+
+- The `pack_bytes` method reads all the bytes from `slice`. This method must
+  fail if the packer does not have enough space to fulfill the request.
+
+Both traits allows us to abstract away any IO operation without relying on
+`std::io`. This has the additional benefit of allowing us to pack and unpack
+from different kinds of buffers without making specific assumptions about the
+implementation of the serialization itself.
+
+## The `Packable` trait
+
+The new `Packable` trait is the following
+
+```rust
+pub trait Packable: Sized {
+    type Error;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
+
+    fn packed_len(&self) -> usize;
+}
+```
+
+- The `Error` associated type represents a semantic error.
+
+- The `pack` method serializes the current value using a `Packer` to write the
+  bytes.
+
+- The `unpack` method deserializes a value using an `Unpacker` to read the
+  bytes.
+
+- The `packed_len` method returns the length in bytes of the serialized value.
```

Comment:
Is there room for `packed_len` to possibly return `None` should the type not be capable of discerning its own packed length?

Reply:
Do you have any examples where that happens?

Reply:
I actually was thinking in writing a "dummy" packer `CountPacker(usize)` that increases everytime a byte is written to it so we don't have to implement `packed_len`. But I don't know if the compiler will be smart enough to inline everything.